test/built-ins/TypedArrays/internals/DefineOwnProperty/detached-buffer.js

```diff
@@ -0,0 +1,124 @@
+// Copyright (C) 2016 the V8 project authors. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: sec-integer-indexed-exotic-objects-defineownproperty-p-desc
+description: |
+  Throws a TypeError if object has valid numeric index and a detached buffer
+info: >
+  9.4.5.3 [[DefineOwnProperty]] ( P, Desc)
+  ...
+  3. If Type(P) is String, then
+    a. Let numericIndex be ! CanonicalNumericIndexString(P).
+    b. If numericIndex is not undefined, then
+      ...
+      xi. If Desc has a [[Value]] field, then
+        1. Let value be Desc.[[Value]].
+        2. Return ? IntegerIndexedElementSet(O, intIndex, value).
+  ...
+
+  9.4.5.9 IntegerIndexedElementSet ( O, index, value )
+
+  ...
+  4. Let buffer be the value of O's [[ViewedArrayBuffer]] internal slot.
+  5. If IsDetachedBuffer(buffer) is true, throw a TypeError exception.
+  ...
+includes: [testTypedArray.js, detachArrayBuffer.js]
+features: [Reflect]
+---*/
+
+var desc = {
+  value: 0,
+  configurable: false,
+  enumerable: true,
+  writable: true
+};
+
+testWithTypedArrayConstructors(function(TA) {
+  var sample = new TA([42]);
+  $DETACHBUFFER(sample.buffer);
+
+  assert.throws(TypeError, function() {
+    Reflect.defineProperty(sample, "0", desc);
+  }, "Throws TypeError on valid numeric index if instance has a detached buffer");
+
+  assert.sameValue(
+    Reflect.defineProperty(sample, "-1", desc),
+    false,
+    "Return false before Detached Buffer check when value is a negative number"
+  );
+
+  assert.sameValue(
+    Reflect.defineProperty(sample, "1.1", desc),
+    false,
+    "Return false before Detached Buffer check when value is not an integer"
+  );
+
+  assert.sameValue(
+    Reflect.defineProperty(sample, "-0", desc),
+    false,
+    "Return false before Detached Buffer check when value is -0"
+  );
+
+  assert.sameValue(
+    Reflect.defineProperty(sample, "0", {
+      configurable: true
```

Comment:
it returns `true` if `value` field is not present, before the check for the detachedBuffer:

```
9.4.5.3 [[DefineOwnProperty]] ( P, Desc)
...
xi. If Desc has a [[Value]] field, then
    1. Let value be Desc.[[Value]].
    2. Return ? IntegerIndexedElementSet(O, intIndex, value).
xii. Return true.
```

I've set a value field to throw a custom error in case it's evaluated before with ToNumber.